### PR TITLE
Port SymSGD trainer

### LINKED ISSUE #623: Port SymSGD
This changes adds parallel Stochastic gradient descent trainer know as SymSGD(https://arxiv.org/abs/1705.08030)

### PR DESCRIPTION
**Pull request by codemzs**
This change adds parallel SGD trainer but disables its multi-threading capabilities because of the lack of OpenMP support by the Clang compiler on linux and macOS build systems and MKL library. It also Supersedes PR #556 by squashing all the commits in one and rebasing with master.

fixes #623

### CODE REVIEW

File: src/Native/build.sh [hunk 1]
```diff
@@ -98,3 +98,7 @@ set -x # turn on trace
 cmake "$DIR" -G "Unix Makefiles" $__cmake_defines
 set +x # turn off trace
 make install
+if [[ "$OSTYPE" == "darwin"* ]]; then
+    echo "Changing libMklImports.dylib's executable path within libSymSgdNative.dylib so that loader can find it."
+    install_name_tool "-change" @loader_path/libMklImports.dylib @rpath/libMklImports.dylib "$RootRepo"/bin/x64."$__configuration"/Native/libSymSgdNative.dylib
```
Comment (eerhardt) [inline src/Native/build.sh]: How is this going to work on an end-user's machine, when they use our NuGet packages?

Comment (eerhardt) [reply src/Native/build.sh]: Note: you can do this stuff inside of CMake too. It may make more sense there, since you won't have to do things like `"$RootRepo"/bin/x64."$__configuration"/Native`.  Plus then it can also be localized in the `SymSgdNative\CmakeLists.txt`, and this high-level build script doesn't need to know about individual assemblies we are building.

Here's an example:  [https://github.com/dotnet/corefx/blob/20bcc468b12119e12428cfd9cce50ef5d5b9cc13/src/Native/System.Security.Cryptography.Native/CMakeLists.txt#L56-L70](https://github.com/dotnet/corefx/blob/20bcc468b12119e12428cfd9cce50ef5d5b9cc13/src/Native/System.Security.Cryptography.Native/CMakeLists.txt#L56-L70)

---
In reply to: [206888376](https://github.com/dotnet/machinelearning/pull/624#discussion_r206888376) [](ancestors = 206888376)

File: test/Microsoft.ML.Predictor.Tests/TestPredictors.cs [hunk 2]
```diff
@@ -231,6 +231,18 @@ public void BinaryClassifierLogisticRegressionTest()
             Done();
         }
 
+        [Fact]
+        [TestCategory("Binary")]
+        public void BinaryClassifierSymSgdTest()
+        {
+            //Results sometimes go out of error tolerance on OS X.
```
Comment (eerhardt) [inline test/Microsoft.ML.Predictor.Tests/TestPredictors.cs]: Can we open an issue to enable this test on macOS?

File: test/Microsoft.ML.TestFramework/Microsoft.ML.TestFramework.csproj [hunk 3]
```diff
@@ -11,6 +11,7 @@
     <ProjectReference Include="..\..\src\Microsoft.ML.FastTree\Microsoft.ML.FastTree.csproj" />
     <ProjectReference Include="..\..\src\Microsoft.ML.Maml\Microsoft.ML.Maml.csproj" />
     <ProjectReference Include="..\..\src\Microsoft.ML.ResultProcessor\Microsoft.ML.ResultProcessor.csproj" />
+    <ProjectReference Include="..\..\src\Microsoft.ML.StandardLearners\Microsoft.ML.StandardLearners.csproj" />
```
Comment (eerhardt) [inline test/Microsoft.ML.TestFramework/Microsoft.ML.TestFramework.csproj]: Is this change necessary now that SymSGD isn't in the StandardLearners project?

File: src/Microsoft.ML/CSharpApi.cs [hunk 4]
```diff
@@ -9761,6 +9773,128 @@ public StochasticGradientDescentBinaryClassifierPipelineStep(Output output)
...
+            /// <summary>
+            /// Number of passes over the data.
+            /// </summary>
+            [TlcModule.SweepableDiscreteParamAttribute("NumberOfIterations", new object[]{1, 5, 10, 20, 30, 40, 50})]
+            public int NumberOfIterations { get; set; } = 50;
+
+            /// <summary>
+            /// Tolerance for difference in average loss in consecutive passes.
+            /// </summary>
+            public float Tol { get; set; } = 0.0001f;
```
Comment (eerhardt) [inline src/Microsoft.ML/CSharpApi.cs]: This should probably be a complete word: `Tolerance`.

File: src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs [hunk 5]
```diff
@@ -0,0 +1,850 @@
...
+                "Multi-threading is not supported currently.", ShortName = "nt")]
+            public int? NumberOfThreads;
+
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Number of passes over the data.", ShortName = "iter", SortOrder = 50)]
+            [TGUI(SuggestedSweeps = "1,5,10,20,30,40,50")]
+            [TlcModule.SweepableDiscreteParam("NumberOfIterations", new object[] { 1, 5, 10, 20, 30, 40, 50 })]
+            public int NumberOfIterations = 50;
+
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Tolerance for difference in average loss in consecutive passes.", ShortName = "tol")]
+            public float Tol = 1e-4f;
```
Comment (eerhardt) [inline src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs]: `Tol` => `Tolerance`?

File: src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs [hunk 6]
```diff
@@ -0,0 +1,850 @@
...
+        private long UsedMemory { get; set; }
+
+        private static unsafe class Native
+        {
+            internal const string DllName = "SymSgdNative";
+            internal const string MklDllName = "MklImports";
+
+            static Native()
+            {
+                if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
```
Comment (eerhardt) [inline src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs]: It may be nice to comment why this is only needed on Windows, and what makes it work on the other platforms.

File: src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs [hunk 7]
```diff
@@ -0,0 +1,850 @@
...
+            {
+                DeallocateSequentially((State*)stateGCHandle.AddrOfPinnedObject());
+            }
+        }
+
+        /// <summary>
+        /// This is the state of a SymSGD learner that is shared between the managed and native code.
+        /// </summary>
+        [StructLayout(LayoutKind.Explicit)]
+        public unsafe struct State
```
Comment (eerhardt) [inline src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs]: Does this need to be public?  I think it should be able to be made internal.

File: src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs [hunk 8]
```diff
@@ -0,0 +1,850 @@
...
+        {
+#pragma warning disable 649 // never assigned
+            [FieldOffset(0x00)]
```
Comment (eerhardt) [inline src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs]: Why the need to explicitly list the offsets here?  This forces us to never work on non-64-bit platforms.

Instead, we should be able to use `[StructLayout(Sequential)]` and then just list the fields as necessary.

For example:  [https://github.com/dotnet/corefx/blob/ba496ae6a9ae055708b630b3f73395a76c2ce6ce/src/System.Drawing.Common/src/System/Drawing/Internal/GpPathData.cs#L9-L15](https://github.com/dotnet/corefx/blob/ba496ae6a9ae055708b630b3f73395a76c2ce6ce/src/System.Drawing.Common/src/System/Drawing/Internal/GpPathData.cs#L9-L15)

File: src/Native/SymSgdNative/SymSgdNative.cpp [hunk 9]
```diff
@@ -0,0 +1,440 @@
...
+#endif
+#include <unordered_map>
+#include "../Stdafx.h"
+#include "Macros.h"
+#include "SparseBLAS.h"
+#include "SymSgdNative.h"
+
+// This method learns for a single instance
+inline void LearnInstance(int instSize, int* instIndices, float* instValues,
+	float label, float alpha, float l2Const, float piw, float& weightScaling, float* weightVector, float& bias) {
```
Comment (eerhardt) [inline src/Native/SymSgdNative/SymSgdNative.cpp]: (nit) It would be nice to follow the coding style already established in the repo and put the opening `{` on a new line:

```C++
inline void LearnInstance(int instSize, int* instIndices, float* instValues,
 float label, float alpha, float l2Const, float piw, float& weightScaling, float* weightVector, float& bias)
{
    //code
}
```

File: src/Native/SymSgdNative/SymSgdNative.cpp [hunk 10]
```diff
@@ -0,0 +1,440 @@
...
+			if (overallAverageLoss < tolerance)
+				break;
+
+			// For shuffling in the next passes, instead of curPermMultiplier, use curPermMultiplier^2 which has exactly the same effect
+			if (needShuffle)
+				curPermMultiplier = (((int64_t)curPermMultiplier * (int64_t)curPermMultiplier) % (int64_t)totalNumInstances);
+			state->PassIteration++;
+		}
+	} else {
+#if defined(USE_OMP)
```
Comment (eerhardt) [inline src/Native/SymSgdNative/SymSgdNative.cpp]: What happens in this case if someone passes in `numThreads > 1`?  Does it just no op?

File: src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs [hunk 11]
```diff
@@ -0,0 +1,850 @@
...
+
+        private TPredictor TrainCore(IChannel ch, RoleMappedData data, LinearPredictor predictor, int weightSetCount)
+        {
+            int numFeatures = data.Schema.Feature.Type.VectorSize;
+            var cursorFactory = new FloatLabelCursor.Factory(data, CursOpt.Label | CursOpt.Features | CursOpt.Weight);
+            int numThreads = 1;
+            ch.CheckUserArg(numThreads > 0, nameof(_args.NumberOfThreads),
+                "The number of threads must be either null or a positive integer.");
+
+            ch.Assert(numThreads > 0);
```
Comment (eerhardt) [inline src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs]: Is this redundant with the line just before it?

File: src/Microsoft.ML.HalLearners/Microsoft.ML.HalLearners.csproj [hunk 12]
```diff
@@ -1,8 +1,9 @@
-﻿<Project Sdk="Microsoft.NET.Sdk">
+<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
     <TargetFramework>netstandard2.0</TargetFramework>
     <IncludeInPackage>Microsoft.ML.HalLearners</IncludeInPackage>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
```
Comment (sfilipi) [inline src/Microsoft.ML.HalLearners/Microsoft.ML.HalLearners.csproj]: > <AllowUnsafeBlocks>true</AllowUnsafeBlocks> [](start = 3, length = 44)

think i removed this

File: src/Native/build.sh (hunk 1 above)
Comment (sfilipi) [inline src/Native/build.sh]: >install_name_tool "-change" @loader_path/libMklImports.dylib @rpath/libMklImports.dylib [](start = 4, length = 87)

I'll do this first part directly to the library, after we produce it. We can remove it from the script later.

File: src/Native/SymSgdNative/SymSgdNative.cpp [hunk 13]
```diff
@@ -0,0 +1,440 @@
...
+
+				// Check if we need to reweight the weight vector
+				if (l2Const > 0.0f) {
+					#pragma omp barrier
+					if (weightScaling < 1e-6) {
+						#pragma omp for
+						for (int featIndex = 0; featIndex < numFeat; featIndex++)
+							weightVector[featIndex] *= weightScaling;
+						if (threadId == 0)
+							weightScaling = 1.0f;
```
Comment (sfilipi) [inline src/Native/SymSgdNative/SymSgdNative.cpp]: >						 [](start = 0, length = 6)

tabs in this file.

File: test/Microsoft.ML.Predictor.Tests/TestPredictors.cs (hunk 2 above)
Comment (codemzs) [reply test/Microsoft.ML.Predictor.Tests/TestPredictors.cs]: We can but linear learners tend to differ on results.

Comment (eerhardt) [reply test/Microsoft.ML.Predictor.Tests/TestPredictors.cs]: At the end of the day, we still need test coverage even if the results differ.  Just because running on a different OS produces different results/behavior doesn't mean we should test on that OS.

Comment (codemzs) [reply test/Microsoft.ML.Predictor.Tests/TestPredictors.cs]: The other test of SymSGD is enabled on macOS too. I can enable this as well just roughly half the time it will fail because of error tolerance.

Comment (eerhardt) [reply test/Microsoft.ML.Predictor.Tests/TestPredictors.cs]: Just open an issue saying that we should enable this test on macOS.  Then when we work on that issue, we can make it stable as well.

File: src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs (hunk 8 above)
Comment (codemzs) [reply src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs]: Given this is a port  it is probably out of scope to make this kind of change without also involving the authors of sysmsgd, instead I will open an issue and assign it to them. Does that sound good?

Comment (eerhardt) [reply src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs]: That works for me.

File: test/Microsoft.ML.Predictor.Tests/TestPredictors.cs (hunk 2 above)
Comment (codemzs) [reply test/Microsoft.ML.Predictor.Tests/TestPredictors.cs]: Will do.

---
In reply to: [206948457](https://github.com/dotnet/machinelearning/pull/624#discussion_r206948457) [](ancestors = 206948457)

File: src/Microsoft.ML.HalLearners/Microsoft.ML.HalLearners.csproj (hunk 12 above)
Comment (codemzs) [reply src/Microsoft.ML.HalLearners/Microsoft.ML.HalLearners.csproj]: I need it for symsgd.

---
In reply to: [206926582](https://github.com/dotnet/machinelearning/pull/624#discussion_r206926582) [](ancestors = 206926582)

File: src/Native/SymSgdNative/SymSgdNative.cpp (hunk 10 above)
Comment (codemzs) [reply src/Native/SymSgdNative/SymSgdNative.cpp]: yep. We have defined as part of arguments that multi-threaded is not supported. 

---
In reply to: [206914794](https://github.com/dotnet/machinelearning/pull/624#discussion_r206914794) [](ancestors = 206914794)

File: src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs [hunk 14]
```diff
@@ -0,0 +1,850 @@
...
+                _instIndices.Dispose();
+                _instValues.Dispose();
+            }
+        }
+
+        private TPredictor TrainCore(IChannel ch, RoleMappedData data, LinearPredictor predictor, int weightSetCount)
+        {
+            int numFeatures = data.Schema.Feature.Type.VectorSize;
+            var cursorFactory = new FloatLabelCursor.Factory(data, CursOpt.Label | CursOpt.Features | CursOpt.Weight);
+            int numThreads = 1;
```
[the remark this answers is not in the document]
Comment (codemzs) [reply src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs]: No, I will open it. We will need to enable this once openmp support comes in.

---
In reply to: [206964865](https://github.com/dotnet/machinelearning/pull/624#discussion_r206964865) [](ancestors = 206964865)

File: src/Native/SymSgdNative/CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+    list(APPEND SOURCES ${VERSION_FILE_PATH})
+    if(CMAKE_SYSTEM_NAME STREQUAL Darwin)
+        message("Linking SymSgdNative with MKL on macOS.")
+        find_library(MKL_LIBRARY libMklImports.dylib HINTS "${CMAKE_SOURCE_DIR}/../../packages/mlnetmkldeps/0.0.0.5/runtimes/osx-x64/native")
+    else()
+        message("Linking SymSgdNative with MKL on linux.")
+        find_library(MKL_LIBRARY libMklImports.so HINTS ${CMAKE_SOURCE_DIR}/../../packages/mlnetmkldeps/0.0.0.5/runtimes/linux-x64/native)
+        SET(CMAKE_SKIP_BUILD_RPATH  FALSE)
+        SET(CMAKE_BUILD_WITH_INSTALL_RPATH FALSE)
+        SET(CMAKE_INSTALL_RPATH "${CMAKE_SOURCE_DIR}/../../packages/mlnetmkldeps/0.0.0.5/runtimes")
```
Comment (eerhardt) [inline src/Native/SymSgdNative/CMakeLists.txt]: I'm concerned that this may work for unit tests, but it may not work on an end-user's machine.  We will have to test this as an end user to verify.

File: src/Native/build.proj [hunk 16]
```diff
@@ -90,6 +90,8 @@
                           RelativePath="Microsoft.ML\runtimes\$(PackageRid)\native" />
       <NativePackageAsset Include="$(NativeAssetsBuiltPath)\$(NativeLibPrefix)FactorizationMachineNative$(NativeLibExtension)"
                           RelativePath="Microsoft.ML\runtimes\$(PackageRid)\native" />
+      <NativePackageAsset Include="$(NativeAssetsBuiltPath)\$(NativeLibPrefix)SymSgdNative$(NativeLibExtension)"
+                          RelativePath="Microsoft.ML\runtimes\$(PackageRid)\native" />
```
Comment (eerhardt) [inline src/Native/build.proj]: This native binary shouldn't go in the `Microsoft.ML` nuget package.  Instead, it needs to go in the `Microsoft.ML.HalLearners` nuget package.

File: src/Native/SymSgdNative/CMakeLists.txt (hunk 15 above)
Comment (codemzs) [reply src/Native/SymSgdNative/CMakeLists.txt]: yep. I will make sure to do this as part of bug bash.

---
In reply to: [206967492](https://github.com/dotnet/machinelearning/pull/624#discussion_r206967492) [](ancestors = 206967492)

File: src/Native/build.proj [hunk 17]
```diff
@@ -90,6 +90,8 @@
                           RelativePath="Microsoft.ML\runtimes\$(PackageRid)\native" />
       <NativePackageAsset Include="$(NativeAssetsBuiltPath)\$(NativeLibPrefix)FactorizationMachineNative$(NativeLibExtension)"
                           RelativePath="Microsoft.ML\runtimes\$(PackageRid)\native" />
+      <NativePackageAsset Include="$(NativeAssetsBuiltPath)\$(NativeLibPrefix)SymSgdNative$(NativeLibExtension)"
+                          RelativePath="Microsoft.Hal\runtimes\$(PackageRid)\native" />
```
Comment (eerhardt) [inline src/Native/build.proj]: The real name here is: `Microsoft.ML.HalLearners`

Comment (codemzs) [reply src/Native/build.proj]: It was on accident.

---
In reply to: [206970539](https://github.com/dotnet/machinelearning/pull/624#discussion_r206970539) [](ancestors = 206970539)

File: src/Native/build.sh (hunk 1 above)
Comment (codemzs) [reply src/Native/build.sh]: We spoke about it offline and the insall_name_tool step can moved out of the script and be done when nuget is created. I'll test the nuget as end user during bug-bash. 

---
In reply to: [206964367](https://github.com/dotnet/machinelearning/pull/624#discussion_r206964367) [](ancestors = 206964367,206888376)
